Code/CPP/TigerVNC/unix/x0vncserver/x0vncserver.man
.TH X0VNCSERVER 1 "Dec 23, 2008" "TigerVNC" "TigerVNC Manual"
.SH NAME
x0vncserver \- TigerVNC Server for X displays
.SH SYNOPSIS
.B x0vncserver
.RI [ options ]
.br
.B x0vncserver -version
.SH DESCRIPTION
.B x0vncserver
is a TigerVNC Server which makes any X display remotely accessible via VNC,
TigerVNC or compatible viewers.  Unlike \fBXvnc\fP(1), it does not create a
virtual display.  Instead, it just shares an existing X server (typically,
that one connected to the physical screen).
.SH OPTIONS
.B x0vncserver
interprets the command line as a list of parameters with optional values.
Running \fBx0vncserver -h\fP will show a list of all valid parameters with
short descriptions.  All parameters are optional, but normally you would have
to use the \fBPasswordFile\fP parameter (see its description below).
.PP
There are several forms of specifying parameters in the command line (here we
use `\fISomeParameter\fP' as an example parameter name):
.TP
.B -\fISomeParameter\fP
Enable the parameter, turn the feature on.  This form can be used with
parameters that simply enable or disable some feature.
.TP
.B -\fISomeParameter\fP=0
Disable the parameter, turn the feature off.
.TP
.B -\fISomeParameter\fP=\fIvalue\fP
Assign the specified \fIvalue\fP to the parameter.  The leading dash can be
omitted, or it can be doubled if desired (like in GNU-style long options).
.PP
Parameter names are case-insensitive, their order in the command line can be
arbitrary.
.SH PARAMETERS
.TP
.B display
The X display name.  If not specified, it defaults to the value of the
DISPLAY environment variable.
.TP
.B rfbport
TCP port to listen for incoming VNC connections (RFB protocol).  The default
port is 5900.
.TP
.B Log
Specifies which log output should be directed to which target logger, and the
level of output to log.  Format is \fIlog\fP:\fItarget\fP:\fIlevel\fP.
Default is \fB*:stderr:30\fP (log everything to stderr, set log level to 30).
Log level should be a value between 0 and 100, higher levels produce more
output.
.TP
.B HostsFile
This parameter allows to specify a file name with IP access control rules.
The file should include one rule per line, and the rule format is one of the
following: +\fIaddress\fP/\fInetmask\fP (accept connections from the
specified address group), -\fIaddress\fP/\fInetmask\fP (reject connections)
or ?\fIaddress\fP/\fInetmask\fP (query the local user).  The first rule
matching the IP address determines the action to be performed.  Rules that
include only an action sign (+, - or ?) will match any IP address.
\fINetmask\fP is optional and can be specified either in dotted format
(e.g. /255.255.255.0), or as a single number of bits (e.g. /24).  Default is
to accept connections from any IP address.
.TP
.B SecurityTypes
Specify which security scheme to use for incoming connections.  Valid values
are \fBNone\fP and \fBVncAuth\fP.  Default is \fBVncAuth\fP.
.TP
.B PasswordFile
Password file for VNC authentication.  There is no default, you should
specify the password file explicitly.  Password file should be created with
the \fBvncpasswd\fP(1) utility.
.TP
.B Password
Obfuscated binary encoding of the password which clients must supply to
access the server.  Using this parameter is insecure, use \fBPasswordFile\fP
parameter instead.
.TP
.B BlacklistThreshold
The number of unauthenticated connection attempts allowed from any individual
host before that host is black-listed.  Default is 5.
.TP
.B BlacklistTimeout
The initial timeout applied when a host is first black-listed.  The host
cannot re-attempt a connection until the timeout expires.  Default is 10.
.TP
.B QueryConnect
Prompt the local user to accept or reject incoming connections.  Default is
off.
.TP
.B QueryConnectTimeout
Number of seconds to show the Accept Connection dialog before rejecting the
connection.  Default is 10.
.TP
.B AlwaysShared
Always treat incoming connections as shared, regardless of the
client-specified setting.  Default is off.
.TP
.B NeverShared
Never treat incoming connections as shared, regardless of the
client-specified setting.  Default is off.
.TP
.B DisconnectClients
Disconnect existing clients if an incoming connection is non-shared.  If
combined with \fBNeverShared\fP then new connections will be refused while
there is a client active.  Default is on.
.TP
.B AcceptKeyEvents
Accept key press and release events from clients.  Default is on.
.TP
.B AcceptPointerEvents
Accept pointer events from clients.  Default is on.
.TP
.B RemapKeys
Comma-separated list of incoming keysyms to remap.  Mappings are expressed as
two hex values, prefixed by \fB0x\fP, and separated by \fB->\fP (`dash' and
`greater than' characters).
.TP
.B Protocol3.3
Always use RFB protocol version 3.3 for backwards compatibility with
badly-behaved clients.  Default is off.
.TP
.B Geometry
This option specifies the screen area that will be shown to VNC clients.  The
format is
.B \fIwidth\fPx\fIheight\fP+\fIxoffset\fP+\fIyoffset\fP
, where `+' signs can be replaced with `\-' signs to specify offsets from the
right and/or from the bottom of the screen.  Offsets are optional, +0+0 is
assumed by default (top left corner).  If the argument is empty, full screen
is shown to VNC clients (this is the default).
.TP
.B MaxProcessorUsage
Maximum percentage of CPU time to be consumed when polling the
screen.  Default is 35.
.TP
.B PollingCycle
Milliseconds per one polling cycle.  Actual interval may be dynamically
adjusted to satisfy \fBMaxProcessorUsage\fP setting.  Default is 30.
.TP
.B VideoArea
This option specifies the screen area that will be handled as video.  The
format is
.B \fIwidth\fPx\fIheight\fP+\fIxoffset\fP+\fIyoffset\fP
, where `+' signs can be replaced with `\-' signs to specify offsets from the
right and/or from the bottom of the screen.  Offsets are optional, +0+0 is
assumed by default (top left corner).  Video area is not checked for actual
changes, it is assumed to be changing continuously and is sent in each screen
update.  Normally, video area is compressed with JPEG (if allowed by the
client).  Note that this setting can be overriden by TigerVNC clients
supporting VideoRectangleSelection protocol extension.  In other words,
clients may choose video selection themselves, but if they do not, then the
\fBVideoArea\fP parameter will take effect.  If the argument is empty, no
video area is set (this is the default).
.TP
.B VideoPriority
Specify the priority of sending video updates.  \fBx0vncserver\fP can be
instructed to treat certain rectangular part of the screen as a video area
and handle it in a special way for improved performance (see documentation on
the \fBVideoArea\fP parameter).  \fBVideoPriority\fP value controls how often
video area will be sent to clients as compared to the rest of the screen.
The priority must be an integer between 0 and 8, and the default value is 2.

\fBVideoPriority\fP set to 1 gives the same priority both to video and to
other pixels.  Higher values give more priority to video.  For example, the
value 5 specifies that the rate of sending video will be five times higher
than the rate of updating the rest of the screen.  If \fBVideoPriority\fP is
set to 0, it gives equal priority to video and other updates (just like the
value 1) and also disables special encoding for video data.  In other words,
\fBVideoPriority\fP video area will be sent as a part of other screen
contents.
.TP
.B CompareFB
Perform pixel comparison on framebuffer to reduce unnecessary updates.
Default is on.
.TP
.B UseSHM
Use MIT-SHM extension if available.  Using that extension accelerates reading
the screen.  Default is on.
.TP
.B OverlayMode
Use overlay mode in IRIX or Solaris (does not have effect in other systems).
This enables system-specific access to complete full-color version of the
screen (the default X visual often provides 256 colors).  Also, in overlay
mode, \fBx0vncserver\fP can show correct mouse cursor.  Default is on.
.TP
.B UseHardwareJPEG
Use hardware-accelerated JPEG compressor for video if available.
\fBx0vncserver\fP can be instructed to treat certain rectangular part of the
screen as a video area and handle it in a special way for improved
performance.  If the client supports Tight encoding and JPEG compression,
such video areas will be sent as JPEG-encoded rectangles.  And if this option
is on, compression will be hardware-accelerated (currently, supported only in
SGI/IRIX equipped with appropriate hardware).  Default is on.
.TP
.B ZlibLevel
Zlib compression level for ZRLE encoding (it does not affect Tight encoding).
Acceptable values are between 0 and 9.  Default is to use the standard
compression level provided by the \fBzlib\fP(3) compression library.
.TP
.B ImprovedHextile
Use improved compression algorithm for Hextile encoding which achieves better
compression ratios by the cost of using slightly more CPU time.  Default is
on.
.TP
.B IdleTimeout
The number of seconds after which an idle VNC connection will be dropped
(zero means no timeout).  Default is 0.
.TP
.B MaxDisconnectionTime
Terminate when no client has been connected for \fIN\fP seconds.  Default is
0.
.TP
.B MaxConnectionTime
Terminate when a client has been connected for \fIN\fP seconds.  Default is
0.
.TP
.B MaxIdleTime
Terminate after \fIN\fP seconds of user inactivity.  Default is 0.
.TP
.B ClientWaitTimeMillis
The number of milliseconds to wait for a client which is no longer
responding.  Default is 20000.
.SH SEE ALSO
.BR Xvnc (1),
.BR vncpasswd (1),
.br
http://www.tigervnc.org/
.SH AUTHOR
This manual page was written by Constantin Kaplinsky.


Code/CPP/TigerVNC/unix/x0vncserver/x0vncserver.cxx
/* Copyright (C) 2002-2005 RealVNC Ltd.  All Rights Reserved.
 * Copyright (C) 2004-2008 Constantin Kaplinsky.  All Rights Reserved.
 *    
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
 * USA.
 */

// FIXME: Check cases when screen width/height is not a multiply of 32.
//        e.g. 800x600.

#include <strings.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>
#include <rfb/Logger_stdio.h>
#include <rfb/LogWriter.h>
#include <rfb/VNCServerST.h>
#include <rfb/Configuration.h>
#include <rfb/SSecurityFactoryStandard.h>
#include <rfb/Timer.h>
#include <network/TcpSocket.h>
#include <tx/TXWindow.h>

#include <vncconfig/QueryConnectDialog.h>

#include <signal.h>
#include <X11/X.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#ifdef HAVE_XTEST
#include <X11/extensions/XTest.h>
#endif

#include <x0vncserver/Geometry.h>
#include <x0vncserver/Image.h>
#include <x0vncserver/XPixelBuffer.h>
#include <x0vncserver/PollingScheduler.h>

// XXX Lynx/OS 2.3: protos for select(), bzero()
#ifdef Lynx
#include <sys/proto.h>
#endif

extern char buildtime[];

using namespace rfb;
using namespace network;

static LogWriter vlog("Main");

IntParameter pollingCycle("PollingCycle", "Milliseconds per one polling "
                          "cycle; actual interval may be dynamically "
                          "adjusted to satisfy MaxProcessorUsage setting", 30);
IntParameter maxProcessorUsage("MaxProcessorUsage", "Maximum percentage of "
                               "CPU time to be consumed", 35);
BoolParameter useShm("UseSHM", "Use MIT-SHM extension if available", true);
BoolParameter useOverlay("OverlayMode", "Use overlay mode under "
                         "IRIX or Solaris", true);
StringParameter displayname("display", "The X display", "");
IntParameter rfbport("rfbport", "TCP port to listen for RFB protocol",5900);
IntParameter queryConnectTimeout("QueryConnectTimeout",
                                 "Number of seconds to show the Accept Connection dialog before "
                                 "rejecting the connection",
                                 10);
StringParameter hostsFile("HostsFile", "File with IP access control rules", "");

//
// Allow the main loop terminate itself gracefully on receiving a signal.
//

static bool caughtSignal = false;

static void CleanupSignalHandler(int sig)
{
  caughtSignal = true;
}


class QueryConnHandler : public VNCServerST::QueryConnectionHandler,
                         public QueryResultCallback {
public:
  QueryConnHandler(Display* dpy, VNCServerST* vs)
    : display(dpy), server(vs), queryConnectDialog(0), queryConnectSock(0) {}
  ~QueryConnHandler() { delete queryConnectDialog; }

  // -=- VNCServerST::QueryConnectionHandler interface
  virtual VNCServerST::queryResult queryConnection(network::Socket* sock,
                                                   const char* userName,
                                                   char** reason) {
    if (queryConnectSock) {
      *reason = strDup("Another connection is currently being queried.");
      return VNCServerST::REJECT;
    }
    if (!userName) userName = "(anonymous)";
    queryConnectSock = sock;
    CharArray address(sock->getPeerAddress());
    delete queryConnectDialog;
    queryConnectDialog = new QueryConnectDialog(display, address.buf,
                                                userName, queryConnectTimeout,
                                                this);
    queryConnectDialog->map();
    return VNCServerST::PENDING;
  }

  // -=- QueryResultCallback interface
  virtual void queryApproved() {
    server->approveConnection(queryConnectSock, true, 0);
    queryConnectSock = 0;
  }
  virtual void queryRejected() {
    server->approveConnection(queryConnectSock, false,
                              "Connection rejected by local user");
    queryConnectSock = 0;
  }
private:
  Display* display;
  VNCServerST* server;
  QueryConnectDialog* queryConnectDialog;
  network::Socket* queryConnectSock;
};


class XDesktop : public SDesktop, public ColourMap
{
public:
  XDesktop(Display* dpy_, Geometry *geometry_)
    : dpy(dpy_), geometry(geometry_), pb(0), server(0),
      oldButtonMask(0), haveXtest(false), maxButtons(0), running(false)
  {
#ifdef HAVE_XTEST
    int xtestEventBase;
    int xtestErrorBase;
    int major, minor;

    if (XTestQueryExtension(dpy, &xtestEventBase,
                            &xtestErrorBase, &major, &minor)) {
      XTestGrabControl(dpy, True);
      vlog.info("XTest extension present - version %d.%d",major,minor);
      haveXtest = true;
    } else {
#endif
      vlog.info("XTest extension not present");
      vlog.info("Unable to inject events or display while server is grabbed");
#ifdef HAVE_XTEST
    }
#endif

  }
  virtual ~XDesktop() {
    stop();
  }

  inline void poll() {
    if (pb)
      pb->poll(server);
  }

  // -=- SDesktop interface

  virtual void start(VNCServer* vs) {

    // Determine actual number of buttons of the X pointer device.
    unsigned char btnMap[8];
    int numButtons = XGetPointerMapping(dpy, btnMap, 8);
    maxButtons = (numButtons > 8) ? 8 : numButtons;
    vlog.info("Enabling %d button%s of X pointer device",
              maxButtons, (maxButtons != 1) ? "s" : "");

    // Create an ImageFactory instance for producing Image objects.
    ImageFactory factory((bool)useShm, (bool)useOverlay);

    // Create pixel buffer and provide it to the server object.
    pb = new XPixelBuffer(dpy, factory, geometry->getRect(), this);
    vlog.info("Allocated %s", pb->getImage()->classDesc());

    server = (VNCServerST *)vs;
    server->setPixelBuffer(pb);

    running = true;
  }

  virtual void stop() {
    running = false;

    delete pb;
    pb = 0;
  }

  inline bool isRunning() {
    return running;
  }

  virtual void pointerEvent(const Point& pos, int buttonMask) {
#ifdef HAVE_XTEST
    if (!haveXtest) return;
    XTestFakeMotionEvent(dpy, DefaultScreen(dpy),
                         geometry->offsetLeft() + pos.x,
                         geometry->offsetTop() + pos.y,
                         CurrentTime);
    if (buttonMask != oldButtonMask) {
      for (int i = 0; i < maxButtons; i++) {
	if ((buttonMask ^ oldButtonMask) & (1<<i)) {
          if (buttonMask & (1<<i)) {
            XTestFakeButtonEvent(dpy, i+1, True, CurrentTime);
          } else {
            XTestFakeButtonEvent(dpy, i+1, False, CurrentTime);
          }
        }
      }
    }
    oldButtonMask = buttonMask;
#endif
  }

  virtual void keyEvent(rdr::U32 key, bool down) {
#ifdef HAVE_XTEST
    if (!haveXtest) return;
    int keycode = XKeysymToKeycode(dpy, key);
    if (keycode)
      XTestFakeKeyEvent(dpy, keycode, down, CurrentTime);
#endif
  }

  virtual void clientCutText(const char* str, int len) {
  }

  virtual Point getFbSize() {
    return Point(pb->width(), pb->height());
  }

  // -=- ColourMap callbacks
  virtual void lookup(int index, int* r, int* g, int* b) {
    XColor xc;
    xc.pixel = index;
    if (index < DisplayCells(dpy,DefaultScreen(dpy))) {
      XQueryColor(dpy, DefaultColormap(dpy,DefaultScreen(dpy)), &xc);
    } else {
      xc.red = xc.green = xc.blue = 0;
    }
    *r = xc.red;
    *g = xc.green;
    *b = xc.blue;
  }

protected:
  Display* dpy;
  Geometry* geometry;
  XPixelBuffer* pb;
  VNCServerST* server;
  int oldButtonMask;
  bool haveXtest;
  int maxButtons;
  bool running;
};


class FileTcpFilter : public TcpFilter
{

public:

  FileTcpFilter(const char *fname)
    : TcpFilter("-"), fileName(NULL), lastModTime(0)
  {
    if (fname != NULL)
      fileName = strdup((char *)fname);
  }

  virtual ~FileTcpFilter()
  {
    if (fileName != NULL)
      free(fileName);
  }

  virtual bool verifyConnection(Socket* s)
  {
    if (!reloadRules()) {
      vlog.error("Could not read IP filtering rules: rejecting all clients");
      filter.clear();
      filter.push_back(parsePattern("-"));
      return false;
    }

    return TcpFilter::verifyConnection(s);
  }

protected:

  bool reloadRules()
  {
    if (fileName == NULL)
      return true;

    struct stat st;
    if (stat(fileName, &st) != 0)
      return false;

    if (st.st_mtime != lastModTime) {
      // Actually reload only if the file was modified
      FILE *fp = fopen(fileName, "r");
      if (fp == NULL)
        return false;

      // Remove all the rules from the parent class
      filter.clear();

      // Parse the file contents adding rules to the parent class
      char buf[32];
      while (readLine(buf, 32, fp)) {
        if (buf[0] && strchr("+-?", buf[0])) {
          filter.push_back(parsePattern(buf));
        }
      }

      fclose(fp);
      lastModTime = st.st_mtime;
    }
    return true;
  }

protected:

  char *fileName;
  time_t lastModTime;

private:

  //
  // NOTE: we silently truncate long lines in this function.
  //

  bool readLine(char *buf, int bufSize, FILE *fp)
  {
    if (fp == NULL || buf == NULL || bufSize == 0)
      return false;

    if (fgets(buf, bufSize, fp) == NULL)
      return false;

    char *ptr = strchr(buf, '\n');
    if (ptr != NULL) {
      *ptr = '\0';              // remove newline at the end
    } else {
      if (!feof(fp)) {
        int c;
        do {                    // skip the rest of a long line
          c = getc(fp);
        } while (c != '\n' && c != EOF);
      }
    }
    return true;
  }

};

char* programName;

static void printVersion(FILE *fp)
{
  fprintf(fp, "TigerVNC Server version %s, built %s\n",
          PACKAGE_VERSION, buildtime);
}

static void usage()
{
  printVersion(stderr);
  fprintf(stderr, "\nUsage: %s [<parameters>]\n", programName);
  fprintf(stderr, "       %s --version\n", programName);
  fprintf(stderr,"\n"
          "Parameters can be turned on with -<param> or off with -<param>=0\n"
          "Parameters which take a value can be specified as "
          "-<param> <value>\n"
          "Other valid forms are <param>=<value> -<param>=<value> "
          "--<param>=<value>\n"
          "Parameter names are case-insensitive.  The parameters are:\n\n");
  Configuration::listParams(79, 14);
  exit(1);
}

int main(int argc, char** argv)
{
  initStdIOLoggers();
  LogWriter::setLogParams("*:stderr:30");

  programName = argv[0];
  Display* dpy;

  for (int i = 1; i < argc; i++) {
    if (Configuration::setParam(argv[i]))
      continue;

    if (argv[i][0] == '-') {
      if (i+1 < argc) {
        if (Configuration::setParam(&argv[i][1], argv[i+1])) {
          i++;
          continue;
        }
      }
      if (strcmp(argv[i], "-v") == 0 ||
          strcmp(argv[i], "-version") == 0 ||
          strcmp(argv[i], "--version") == 0) {
        printVersion(stdout);
        return 0;
      }
      usage();
    }

    usage();
  }

  CharArray dpyStr(displayname.getData());
  if (!(dpy = XOpenDisplay(dpyStr.buf[0] ? dpyStr.buf : 0))) {
    // FIXME: Why not vlog.error(...)?
    fprintf(stderr,"%s: unable to open display \"%s\"\r\n",
            programName, XDisplayName(dpyStr.buf));
    exit(1);
  }

  signal(SIGHUP, CleanupSignalHandler);
  signal(SIGINT, CleanupSignalHandler);
  signal(SIGTERM, CleanupSignalHandler);

  try {
    TXWindow::init(dpy,"x0vncserver");
    Geometry geo(DisplayWidth(dpy, DefaultScreen(dpy)),
                 DisplayHeight(dpy, DefaultScreen(dpy)));
    if (geo.getRect().is_empty()) {
      vlog.error("Exiting with error");
      return 1;
    }
    XDesktop desktop(dpy, &geo);

    VNCServerST server("x0vncserver", &desktop);
    QueryConnHandler qcHandler(dpy, &server);
    server.setQueryConnectionHandler(&qcHandler);

    TcpListener listener((int)rfbport);
    vlog.info("Listening on port %d", (int)rfbport);

    const char *hostsData = hostsFile.getData();
    FileTcpFilter fileTcpFilter(hostsData);
    if (strlen(hostsData) != 0)
      listener.setFilter(&fileTcpFilter);
    delete[] hostsData;

    PollingScheduler sched((int)pollingCycle, (int)maxProcessorUsage);

    while (!caughtSignal) {
      struct timeval tv;
      fd_set rfds;
      std::list<Socket*> sockets;
      std::list<Socket*>::iterator i;

      // Process any incoming X events
      TXWindow::handleXEvents(dpy);

      FD_ZERO(&rfds);
      FD_SET(listener.getFd(), &rfds);
      server.getSockets(&sockets);
      int clients_connected = 0;
      for (i = sockets.begin(); i != sockets.end(); i++) {
        if ((*i)->isShutdown()) {
          server.removeSocket(*i);
          delete (*i);
        } else {
          FD_SET((*i)->getFd(), &rfds);
          clients_connected++;
        }
      }

      if (!clients_connected)
        sched.reset();

      if (sched.isRunning()) {
        int wait_ms = sched.millisRemaining();
        if (wait_ms > 500) {
          wait_ms = 500;
        }
        tv.tv_usec = wait_ms * 1000;
#ifdef DEBUG
        // fprintf(stderr, "[%d]\t", wait_ms);
#endif
      } else {
        tv.tv_usec = 100000;
      }
      tv.tv_sec = 0;

      // Do the wait...
      sched.sleepStarted();
      int n = select(FD_SETSIZE, &rfds, 0, 0, &tv);
      sched.sleepFinished();

      if (n < 0) {
        if (errno == EINTR) {
          vlog.debug("Interrupted select() system call");
          continue;
        } else {
          throw rdr::SystemException("select", errno);
        }
      }

      // Accept new VNC connections
      if (FD_ISSET(listener.getFd(), &rfds)) {
        Socket* sock = listener.accept();
        if (sock) {
          server.addSocket(sock);
        } else {
          vlog.status("Client connection rejected");
        }
      }

      Timer::checkTimeouts();
      server.checkTimeouts();

      // Client list could have been changed.
      server.getSockets(&sockets);

      // Nothing more to do if there are no client connections.
      if (sockets.empty())
        continue;

      // Process events on existing VNC connections
      for (i = sockets.begin(); i != sockets.end(); i++) {
        if (FD_ISSET((*i)->getFd(), &rfds))
          server.processSocketEvent(*i);
      }

      if (desktop.isRunning() && sched.goodTimeToPoll()) {
        sched.newPass();
        desktop.poll();
      }
    }

  } catch (rdr::Exception &e) {
    vlog.error(e.str());
    return 1;
  }

  TXWindow::handleXEvents(dpy);

  vlog.info("Terminated");
  return 0;
}
